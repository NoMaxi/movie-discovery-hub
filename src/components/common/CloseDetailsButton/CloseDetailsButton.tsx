import { Link, useSearchParams } from "react-router-dom";

export const CloseDetailsButton = () => {
    const [searchParams] = useSearchParams();
    const searchParamsString = searchParams.toString();

    return (
        <Link
            data-testid="close-details-button"
            to={searchParamsString ? `/?${searchParamsString}` : "/"}
            className="
                absolute top-0 right-0 p-2 z-20 text-5xl text-primary
                hover:text-red-400 cursor-pointer transition-colors
            "
            title="Close movie details"
            aria-label="Close movie details"
        >
            &times;
        </Link>
    );
};
