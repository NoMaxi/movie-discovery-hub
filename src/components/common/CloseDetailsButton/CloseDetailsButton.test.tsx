import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { CloseDetailsButton } from "./CloseDetailsButton";

jest.mock("react-router-dom", () => {
    const actual = jest.requireActual("react-router-dom");
    return {
        ...actual,
        useSearchParams: jest.fn(),
    };
});

const { useSearchParams } = jest.requireMock("react-router-dom");

describe("CloseDetailsButton", () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    const mockSearchParams = (params: string = "") => {
        const mockSearchParamsInstance = new URLSearchParams(params);
        (useSearchParams as jest.Mock).mockReturnValue([
            mockSearchParamsInstance,
            jest.fn(),
        ]);
    };

    const renderButton = (searchParamsString: string = "") => {
        mockSearchParams(searchParamsString);
        return render(
            <MemoryRouter>
                <CloseDetailsButton />
            </MemoryRouter>,
        );
    };

    test("should render the close button with correct styling", () => {
        const { asFragment } = renderButton();
        const closeButton = screen.getByRole("link", { name: /close movie details/i });

        expect(closeButton).toBeInTheDocument();
        expect(closeButton).toHaveClass("absolute");
        expect(closeButton).toHaveClass("top-0");
        expect(closeButton).toHaveClass("right-0");
        expect(closeButton).toHaveClass("text-5xl");
        expect(closeButton).toHaveTextContent("Ã—");
        expect(asFragment()).toMatchSnapshot();
    });

    test("should navigate to root path when no search params are present", () => {
        renderButton();
        const closeButton = screen.getByRole("link", { name: /close movie details/i });

        expect(closeButton).toHaveAttribute("href", "/");
    });

    test("should preserve search params in the navigation link", () => {
        renderButton("genre=action&sort=rating");
        const closeButton = screen.getByRole("link", { name: /close movie details/i });

        expect(closeButton).toHaveAttribute("href", "/?genre=action&sort=rating");
    });

    test("should have proper accessibility attributes", () => {
        renderButton();
        const closeButton = screen.getByRole("link", { name: /close movie details/i });

        expect(closeButton).toHaveAttribute("title", "Close movie details");
        expect(closeButton).toHaveAttribute("aria-label", "Close movie details");
    });
});
