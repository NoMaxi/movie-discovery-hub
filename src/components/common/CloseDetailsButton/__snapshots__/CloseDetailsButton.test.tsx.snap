// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloseDetailsButton should render the close button with correct styling 1`] = `
<DocumentFragment>
  <a
    aria-label="Close movie details"
    class="
                absolute top-0 right-0 p-2 z-20 text-5xl text-[var(--color-primary)]
                hover:text-red-400 cursor-pointer transition-colors
            "
    data-discover="true"
    href="/"
    title="Close movie details"
  >
    Ã—
  </a>
</DocumentFragment>
`;
